// @ros2jsguy/mpu6050-motion-data/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
const rpio = require("rpio");
const three_math_ts_1 = require("three-math-ts");
var Utils;
(function (Utils) {
    function sleep(secs) {
        rpio.sleep(secs);
    }
    Utils.sleep = sleep;
    ;
    function msleep(mills) {
        rpio.msleep(mills);
    }
    Utils.msleep = msleep;
    function usleep(usecs) {
        rpio.usleep(usecs);
    }
    Utils.usleep = usleep;
    function clamp(value, min, max) {
        return three_math_ts_1.MathUtils.clamp(value, min, max);
    }
    Utils.clamp = clamp;
    function toBinaryString(integer, withPaddingLength = 8) {
        const str = integer.toString(2);
        return str.padStart(withPaddingLength, "0");
    }
    Utils.toBinaryString = toBinaryString;
    function toHexString(integer, withPaddingLength = 2) {
        const str = integer.toString(16);
        return str.padStart(withPaddingLength, "0");
    }
    Utils.toHexString = toHexString;
    function degToRad(degrees) {
        return three_math_ts_1.MathUtils.degToRad(degrees);
    }
    Utils.degToRad = degToRad;
    function radToDeg(radians) {
        return three_math_ts_1.MathUtils.radToDeg(radians);
    }
    Utils.radToDeg = radToDeg;
    function celciusToF(celcius) {
        return celcius * 1.8 + 32;
    }
    Utils.celciusToF = celciusToF;
})(Utils = exports.Utils || (exports.Utils = {}));
